File: packages/vm/wasmlib/keys.go
// Copyright 2020 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

package wasmlib

type MapKey interface {
	KeyId() Key32
}

type Key string

func (key Key) KeyId() Key32 {
	return GetKeyIdFromString(string(key))
}

type Key32 int32

func (key Key32) KeyId() Key32 {
	return key
}

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

const (
	KeyAddress         = Key32(-1)
	KeyAggregateBls    = Key32(-2)
	KeyBalances        = Key32(-3)
	KeyBase58Bytes     = Key32(-4)
	KeyBase58String    = Key32(-5)
	KeyCall            = Key32(-6)
	KeyCaller          = Key32(-7)
	KeyChainOwnerId    = Key32(-8)
	KeyColor           = Key32(-9)
	KeyContractCreator = Key32(-10)
	KeyContractId      = Key32(-11)
	KeyDeploy          = Key32(-12)
	KeyEvent           = Key32(-13)
	KeyExports         = Key32(-14)
	KeyHashBlake2b     = Key32(-15)
	KeyHashSha3        = Key32(-16)
	KeyHname           = Key32(-17)
	KeyIncoming        = Key32(-18)
	KeyLength          = Key32(-19)
	KeyLog             = Key32(-20)
	KeyMaps            = Key32(-21)
	KeyName            = Key32(-22)
	KeyPanic           = Key32(-23)
	KeyParams          = Key32(-24)
	KeyPost            = Key32(-25)
	KeyRandom          = Key32(-26)
	KeyResults         = Key32(-27)
	KeyReturn          = Key32(-28)
	KeyState           = Key32(-29)
	KeyTimestamp       = Key32(-30)
	KeyTrace           = Key32(-31)
	KeyTransfers       = Key32(-32)
	KeyUtility         = Key32(-33)
	KeyValid           = Key32(-34)
	KeyValidBls        = Key32(-35)
	KeyValidEd25519    = Key32(-36)
	KeyZzzzzzz         = Key32(-37)
)

File: packages/vm/wasmlib/src/keys.rs
// Copyright 2020 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

use crate::host::*;

// conversion of data bytes to key id
pub trait MapKey {
    fn get_key_id(&self) -> Key32;
}

// implementations for both flavors of Rust string
impl MapKey for str {
    fn get_key_id(&self) -> Key32 {
        get_key_id_from_string(self)
    }
}

impl MapKey for String {
    fn get_key_id(&self) -> Key32 {
        get_key_id_from_string(self)
    }
}

// special type for predefined key ids
#[derive(Clone, Copy)]
pub struct Key32(pub i32);

// implementation for predefined key ids
impl MapKey for Key32 {
    fn get_key_id(&self) -> Key32 {
        *self
    }
}

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

// all predefined key id values should exactly match their counterpart values on the host!
// note that predefined key ids are negative values to distinguish them from indexes

// @formatter:off
pub const KEY_ACCOUNT_ID       : Key32 = Key32(-1);
pub const KEY_ADDRESS          : Key32 = Key32(-2);
pub const KEY_BALANCES         : Key32 = Key32(-3);
pub const KEY_BASE58_DECODE    : Key32 = Key32(-4);
pub const KEY_BASE58_ENCODE    : Key32 = Key32(-5);
pub const KEY_BLS_ADDRESS      : Key32 = Key32(-6);
pub const KEY_BLS_AGGREGATE    : Key32 = Key32(-7);
pub const KEY_BLS_VALID        : Key32 = Key32(-8);
pub const KEY_CALL             : Key32 = Key32(-9);
pub const KEY_CALLER           : Key32 = Key32(-10);
pub const KEY_CHAIN_ID         : Key32 = Key32(-11);
pub const KEY_CHAIN_OWNER_ID   : Key32 = Key32(-12);
pub const KEY_COLOR            : Key32 = Key32(-13);
pub const KEY_CONTRACT         : Key32 = Key32(-14);
pub const KEY_CONTRACT_CREATOR : Key32 = Key32(-15);
pub const KEY_DEPLOY           : Key32 = Key32(-16);
pub const KEY_ED25519_ADDRESS  : Key32 = Key32(-17);
pub const KEY_ED25519_VALID    : Key32 = Key32(-18);
pub const KEY_EVENT            : Key32 = Key32(-19);
pub const KEY_EXPORTS          : Key32 = Key32(-20);
pub const KEY_HASH_BLAKE2B     : Key32 = Key32(-21);
pub const KEY_HASH_SHA3        : Key32 = Key32(-22);
pub const KEY_HNAME            : Key32 = Key32(-23);
pub const KEY_INCOMING         : Key32 = Key32(-24);
pub const KEY_LENGTH           : Key32 = Key32(-25);
pub const KEY_LOG              : Key32 = Key32(-26);
pub const KEY_MAPS             : Key32 = Key32(-27);
pub const KEY_MINTED           : Key32 = Key32(-28);
pub const KEY_PANIC            : Key32 = Key32(-29);
pub const KEY_PARAMS           : Key32 = Key32(-30);
pub const KEY_POST             : Key32 = Key32(-31);
pub const KEY_RANDOM           : Key32 = Key32(-32);
pub const KEY_REQUEST_ID       : Key32 = Key32(-33);
pub const KEY_RESULTS          : Key32 = Key32(-34);
pub const KEY_RETURN           : Key32 = Key32(-35);
pub const KEY_STATE            : Key32 = Key32(-36);
pub const KEY_TIMESTAMP        : Key32 = Key32(-37);
pub const KEY_TRACE            : Key32 = Key32(-38);
pub const KEY_TRANSFERS        : Key32 = Key32(-39);
pub const KEY_UTILITY          : Key32 = Key32(-40);
pub const KEY_ZZZZZZZ          : Key32 = Key32(-41);
// @formatter:on
